common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalShuffleBlockHandler.java

```diff
@@ -209,4 +190,47 @@ private ShuffleMetrics() {
     }
   }
 
+  private class ManagedBufferIterator implements Iterator<ManagedBuffer> {
+
+    private int index = 0;
+    private String appId;
+    private String execId;
+    private String shuffleId;
+    // mapId and reduceId are stored as bytes to save memory.
+    private byte[][] mapIdAndReduceIds;
+
+    ManagedBufferIterator(String appId, String execId, String[] blockIds) {
+      this.appId = appId;
+      this.execId = execId;
+      String blockId = blockIds[0];
+      String[] blockIdParts = blockId.split("_");
+      if (blockIdParts.length < 4) {
+        throw new IllegalArgumentException("Unexpected block id format: " + blockId);
+      } else if (!blockIdParts[0].equals("shuffle")) {
+        throw new IllegalArgumentException("Expected shuffle block id, got: " + blockId);
+      }
+      this.shuffleId = blockIdParts[1];
+      mapIdAndReduceIds = new byte[blockIds.length][];
+      if (blockIds.length > 0) {
+        for (int i = 0; i< blockIds.length; i++) {
+          mapIdAndReduceIds[i] = (blockIdParts[2] + "_" + blockIdParts[3]).getBytes();
```

Comment:
Yes, I think this is a good idea. In current change, I make it to be `int[blockIds.length][2]`. I'm not sure if I understand your comment correctly. Please take another look :)